### ISSUE TITLE
Add support Read+Seek trait

### ISSUE THREAD

**Issue by tsubakisakura** [LaurentMazare/tch-rs#438]
Hello!

I have tried restoring VarStore from ```std::io::File```, ```std::io::Cursor```, etc, but there was no direct loading API.
So, I thought it would be convenient if the following APIs were available in the VarStore.

~~~
pub trait ReadStream : Read+Seek {}

impl VarStore {
    ...
    pub fn load_from_stream<S: ReadStream>(&mut self, stream: S) -> Result<(), TchError>
    ...
}
~~~

libtorch has a ```torch::jit::load``` function that takes ```caffe2::serialize::ReadAdapterInterface```.
If I implemented this, it seemed to connect to the Rust side. I tried to implement it and it worked fine.
How about an API like this? If this API policy is acceptable, I will make a pull request.

Thanks.

**Comment by LaurentMazare** [comment]
Hello,
Sounds like an interesting idea. Could you say a bit more about the potential use case for this, how the api would look like and what would have to be implemented? I just gave a quick look at `caffe2::serialize::ReadAdapterInterface` and it seems very generic so it's a bit unclear to me what aspects you would want to cover here.

**Comment by tsubakisakura** [comment]
Thank you for your reply!

I have already implemented the required implementation code, so I pushed it.
https://github.com/tsubakisakura/tch-rs/commit/f4714da233ab2101174539a52a66b1555a7fa50b

My real purpose is to get the thread to do the job as much as possible.
That's why I wanted to run disk I/O in a separate thread.
The thread does not need disk access if ```std::io::Cursor``` is available.

But I realized there might be another way.
The I/O thread can use ```Tensor::load_multi``` if the VarStore can be initialized with ```Vec<(String,Tensor)>```.
This seems to be more efficient because the deserialization process can be done by the I/O thread.

**Comment by LaurentMazare** [comment]
Thanks, the change looks fairly reasonable, happy to get some PR with it. Two small points:
- Could the `at_load_from_stream_callback_with_device` and `at_load_from_stream_callback` C++ functions be shared, e.g. using a device of -1 to represent the case where no device was provided?
- Could the `extern "C"` bits be moved to `torch_api.h` rather than `torch_api.cpp` (similar to all the other functions)?

**Comment by tsubakisakura** [comment]
Thank you for your feedback and reviews!

Since ```Device::Cpu.c_int()``` is also -1, ```at_load_from_stream_callback``` cannot determine whether the default argument or ```Device::Cpu``` was specified.
So I added a bool argument. It may be a bit verbose, but I think it's safe.

Moved ```extern "C"``` functions to torch_api.h.
In addition, I moved functions and traits from ```tensor.rs``` to the ```torch-sys::io``` module, in order to eliminate the static reverse dependency from torch-sys to tch-rs.
(I added ```io.rs``` because it seemed inappropriate to add it to ```lib.rs```. Was this good?)
